From daecec2cbea7cc28cd9f32388712811f2d0bee92 Mon Sep 17 00:00:00 2001
From: Stephan Gerhold <stephan.gerhold@linaro.org>
Date: Tue, 5 Aug 2025 17:44:32 +0200
Subject: [PATCH] remoteproc: qcom_q6v5_pas: Attach running remoteproc if
 firmware is missing

Initialize the q6v5 attach state and detect running firmware before remoteproc
registration. Unwind the attach helper on failure. Gaokun EL2 uses the
upstream boolean auto_boot policy; it has no reset-capable DSP or lite PAS ID.

Signed-off-by: Stephan Gerhold <stephan.gerhold@linaro.org>
---
diff --git a/drivers/remoteproc/qcom_q6v5_pas.c b/drivers/remoteproc/qcom_q6v5_pas.c
index a005546..327c8f6 100644
--- a/drivers/remoteproc/qcom_q6v5_pas.c
+++ b/drivers/remoteproc/qcom_q6v5_pas.c
@@ -527,9 +527,9 @@ static int qcom_pas_attach(struct rproc *rproc)
 	bool stop_state;
 	int ret;
 
-	pas->q6v5.handover_issued = true;
-
-	pas->q6v5.running = true;
+	ret = qcom_q6v5_attach(&pas->q6v5);
+	if (ret)
+		return ret;
 	ret = irq_get_irqchip_state(pas->q6v5.fatal_irq,
 				    IRQCHIP_STATE_LINE_LEVEL, &crash_state);
 	if (ret)
@@ -573,6 +573,7 @@ unroll_attach:
 	pas->rproc->state = RPROC_OFFLINE;
 	ret = -EINVAL;
 disable_running:
+	qcom_q6v5_unprepare(&pas->q6v5);
 	pas->q6v5.running = false;
 
 	return ret;
@@ -913,6 +914,14 @@ static int qcom_pas_probe(struct platform_device *pdev)
 	if (ret)
 		goto detach_proxy_pds;
 
+	/*
+	 * Unfortunately, the PAS interface does not provide a reliable way to
+	 * check if a certain pas_id is currently running. Reading the smp2p
+	 * state is the best we can do to check if the remoteproc is already
+	 * running during boot.
+	 */
+	qcom_q6v5_read_smp2p_state(&pas->q6v5);
+
 	qcom_add_glink_subdev(rproc, &pas->glink_subdev, desc->ssr_name);
 	qcom_add_smd_subdev(rproc, &pas->smd_subdev);
 	qcom_add_pdm_subdev(rproc, &pas->pdm_subdev);
-- 
2.56.0
